cassiopeia/preprocess/UMI_utils.py

```diff
@@ -322,6 +346,64 @@ def form_clusters(
     return clusters
 
 
+def form_clusters_bayesian(
+    als: List[pysam.AlignedSegment],
+    q_threshold: Optional[int] = None,
+    proportion: float = 0.05,
+) -> List[pysam.AlignedSegment]:
+    """Forms clusters from a list of aligned segments (reads) probabilistically.
+
+    This function has the same purpose as :func:`form_clusters`, but instead
+    of using hardcoded quality thresholds, takes a probabilisitc approach to
+    forming read clusters by treating the quality scores as probabilities (in
+    fact, they are actually encoded probabilities of error). The two optional
+    arguments specify a dynamic threshold to assign a specific sequence to
+    a cluster. Specifically, for a given cluster sequence, a sequence is
+    assigned to this cluster (consensus) if its "match probability" to this
+    consensus is <= max(min(match probability), `q_threshold` * (`proportion` * length of longest sequence))
```

Comment:
this line is too long

Reply:
I wonder how this got through `black`... Perhaps i forgot to run it

Reply:
Sometimes `black` doesn't handle these things well haha. I sometimes put math on a separate line.

Reply:
Huh, seems black doesn't want to line break here. will do manually